### ISSUE TITLE
Biome/code actions freezing

### ISSUE THREAD

**Issue by showme032** [zed-industries/zed#10786]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Sometimes after Biome has made a styling/formatting suggestion, it freezes completely, and code actions become unresponsive, even after the suggestion is fixed. Restarting the editor fixes the issue until the next occurrence.

### Environment

Zed: v0.132.0 (Zed Preview)
OS: macOS 14.4.1
Memory: 16 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

https://github.com/zed-industries/zed/assets/8236625/62a6bc6e-f037-454a-856e-3efc72a4435a

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

[log.log](https://github.com/zed-industries/zed/files/15044312/log.log)

**Comment by JosephTLyons** [comment]
I'm wondering if this has anything to do with multiple language servers running at once. cc @maxbrunsfeld?

**Comment by showme032** [comment]
I thought that might be one of the reasons, but if I understand correctly, atm, eslint can't be disabled?

**Comment by maxdeviant** [comment]
> I thought that might be one of the reasons, but if I understand correctly, atm, eslint can't be disabled?

@showme032 If you're comfortable building Zed from source, you could try commenting out these lines:

https://github.com/zed-industries/zed/blob/2ee257a5621d3c2714945e117d958383c14ae513/crates/languages/src/lib.rs#L196-L246

That will disable the built-in TypeScript and ESLint support.

Then run a debug build of Zed using the Biome extension to see if the that addresses the issue.

**Comment by showme032** [comment]
> > I thought that might be one of the reasons, but if I understand correctly, atm, eslint can't be disabled?
> 
> 
> 
> @showme032 If you're comfortable building Zed from source, you could try commenting out these lines:
> 
> 
> 
> https://github.com/zed-industries/zed/blob/2ee257a5621d3c2714945e117d958383c14ae513/crates/languages/src/lib.rs#L196-L246
> 
> 
> 
> That will disable the built-in TypeScript and ESLint support.
> 
> 
> 
> Then run a debug build of Zed using the Biome extension to see if the that addresses the issue.

I'll give it a go and try it out, once I get some time in the following days.

**Comment by showme032** [comment]
Thanks for the reply. I'll give it a go and try it out, once I get some time in the following days.

**Comment by maxdeviant** [comment]
> Thanks for the reply. I'll give it a go and try it out, once I get some time in the following days.

In the current Zed Preview (v0.133.2) you can try disabling the built-in ESLint language server to see if that fixes the problem.

Add this to your settings and then restart the language servers (or restart Zed entirely, to be on the safe side):

```json
{
  "languages": {
      "JavaScript": {
        "language_servers": ["biome", "!eslint", "..."]
      }
    }
}
```

**Comment by chungweileong94** [comment]
> > Thanks for the reply. I'll give it a go and try it out, once I get some time in the following days.
> 
> In the current Zed Preview (v0.133.2) you can try disabling the built-in ESLint language server to see if that fixes the problem.
> 
> Add this to your settings and then restart the language servers (or restart Zed entirely, to be on the safe side):
> 
> ```json
> {
>   "languages": {
>       "JavaScript": {
>         "language_servers": ["biome", "!eslint", "..."]
>       }
>     }
> }
> ```

For some reason, the eslint LSP always running even with the settings.
![image](https://github.com/zed-industries/zed/assets/15154097/0a8c1d9e-0d16-4008-b059-433f615aa4e8)

**Comment by chungweileong94** [comment]
After coding without `eslint` enabled for a while, and this is what I found:
- Some LSPs still failed randomly after using it some times.
- For instance, my prettier formatter stopped working all of a sudden, while other LSPs like `typescript-language-server` & `biome` are still up and running. After restarting the LSP, sometimes `typescript-language-server` will stop working, and I have to restart the LSP for the second time, or even multiple times to get all LSPs to start working again
- Note that the order of the LSP start failing doesn't really matter, they all fail randomly each time.

Unfortunately, I wasn't able to capture log at the time. Will do so next time.

**Comment by christian-gama** [comment]
With Biome extension enabled, my code actions freezes too. The only way to make it work again is by restarting the whole IDE. I couldn't figure it out what triggers the freezing, but I noticed it freezes in a short period of time after initializing the IDE. Disabling the extension avoid the code actions from freezing. 

Here is what I have in my settings.json for TypeScript.

```json
{
  "language_overrides": {
    "TypeScript": {
      "language_servers": ["biome", "!eslint", "!prettier", "..."],
      "formatter": {
        "external": {
          "command": "./node_modules/.bin/biome",
          "arguments": ["check", "--apply-unsafe", "--stdin-file-path={buffer_path}"]
        }
      }
    }
  }
}
```

**Comment by chungweileong94** [comment]
I recorded a screen recording a while ago that showcases how to reproduce the problem, and I figure it will be easy to reference it here, https://github.com/biomejs/biome-zed/issues/14#issuecomment-2094061529.

Basically, I can consistently reproduce the problem by performing "Go to definitions" multiple times, and at some point, Zed will start to malfunction, like code formatting/code action stop working.

**Comment by kevinfaveri** [comment]
Hey guys, any news on this? TS/JS development seems to happen a lot, so not only Biome seems affected... Really a dealbreaker to migrate off to Zed

**Comment by cmjoseph07** [comment]
Agreed with @kevinfaveri, deal breaker to migrate over. Should be something that is higher priority due to number of TS/JS devs currently.

**Comment by luckydye** [comment]
This may already be fixed, have u tried the latest preview builds? At least I am not seeing this anymore..

**Comment by chungweileong94** [comment]
> This may already be fixed, have u tried the latest preview builds? At least I am not seeing this anymore..

I got the same conclusion, haven't seen it for the past few months.

**Comment by cmjoseph07** [comment]
@luckydye 

Seems to be corrected on latest, thanks for letting me know!
